servlet/src/main/java/com/linecorp/armeria/server/servlet/DefaultServletHttpRequest.java

```diff
@@ -0,0 +1,738 @@
+    private final String requestURI;
+    private final String characterEncoding;
+    private final Locale[] locales;
+
+    @Nullable
+    private final Cookie[] cookies;
+    @Nullable
+    private final String pathInfo;
+    @Nullable
+    private final String queryString;
+
+    //Can't be final because user will decide reader or inputStream is initialize.
+    @Nullable
+    private BufferedReader reader;
+
+    private final Map<String, String[]> unmodifiableParameterMap = new AbstractMap<String, String[]>() {
+        @Override
+        public Set<Entry<String, String[]>> entrySet() {
+            if (isEmpty()) {
+                return Collections.emptySet();
+            }
+            return parameterMap.entrySet()
+                               .stream()
+                               .map(x -> new SimpleImmutableEntry<>(
+                                       x.getKey(),
+                                       x.getValue() != null ? x.getValue().toArray(new String[0]) : null))
+                               .collect(Collectors.toSet());
+        }
+
+        @Override
+        @Nullable
+        public String[] get(@Nullable Object key) {
+            final List<String> value = parameterMap.get(key);
+            if (value == null) {
+                return null;
+            } else {
+                return value.toArray(new String[0]);
+            }
+        }
+
+        @Override
+        public boolean containsKey(Object key) {
+            requireNonNull(key, "key");
+            return parameterMap.containsKey(key);
+        }
+
+        @Override
+        public boolean containsValue(Object value) {
+            requireNonNull(value, "value");
+            return parameterMap.toSingleValueMap().containsValue(value);
+        }
+
+        @Override
+        public int size() {
+            return parameterMap.size();
+        }
+    };
+
+    DefaultServletHttpRequest(ServiceRequestContext serviceRequestContext,
+                              DefaultServletContext servletContext,
+                              AggregatedHttpRequest httpRequest) {
+        requireNonNull(serviceRequestContext, "serviceRequestContext");
+        requireNonNull(servletContext, "servletContext");
+        requireNonNull(httpRequest, "request");
+
+        this.serviceRequestContext = serviceRequestContext;
+        this.servletContext = servletContext;
+        this.httpRequest = httpRequest;
+        inputStream = new DefaultServletInputStream(Unpooled.wrappedBuffer(httpRequest.content().array()));
+
+        final MediaType contentType = httpRequest.headers().contentType();
+        if (contentType != null && contentType.charset() != null) {
+            characterEncoding = contentType.charset().name();
+        } else {
+            characterEncoding = servletContext.getRequestCharacterEncoding();
+        }
+
+        requestURI = decodeRequestURI();
+        decodeUrlParameter();
+        decodeBody();
```

Comment:
Let's remove `decodeUrlParameter` and `decodeBody`.
We can use `QueryParam` for this purpose.
First we need to move this [logic](https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java#L1203) to an internal public class to use together.
```java
package com.linecorp.armeria.internal.server;

public final class QueryParamUtil {
    ...
}
```
Then, we can use the method to parse assign to the field `QueryParams queryParams`.
So we can remove `parameterMap` and `unmodifiableParameterMap`.